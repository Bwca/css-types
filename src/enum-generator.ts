export class EnumGenerator {
  public static generateEnumStringForWritingToFile(enumName: string, enumBody: string): string {
    return `export enum ${enumName} ${enumBody}`;
  }

  public static generateEnumNameFromPath(path: string): string {
    const fileName = path.match(/([^\/\\]*)\.scss$/);
    const enumName = (fileName && fileName[1]) || 'styles';
    return EnumGenerator.generateCamelCaseName(enumName);
  }

  public static generateEnumBody(cssClasses: string[]): string {
    let enumBody = '{\n';

    cssClasses.forEach((i) => {
      const propName = EnumGenerator.generateCamelCaseName(i);
      const noDotClassName = i.replace(/^\./, '');
      enumBody += `  ${propName} = '${noDotClassName}',\n`;
    });

    enumBody += '}\n';
    return enumBody;
  }

  public static generateCamelCaseName(str: string): string {
    if (!/[-_]/g.test(str)) {
      return str.replace(/^\../, ([, firstLetter]) => firstLetter.toUpperCase());
    }
    return str
      .replace(/(\-){2,}/g, '$1')
      .replace(/(_){2,}/g, '$1')
      .replace(/[\.\-_][a-z]/g, ([, letter]) => letter.toUpperCase());
  }
}
