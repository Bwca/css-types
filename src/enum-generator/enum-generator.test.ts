import { EnumGenerator } from './enum-generator';

describe(`${EnumGenerator.name} tests for enum name generation`, () => {
  it('should generate enum name from a provided path to a scss stylesheet', () => {
    // Arrange
    const filePath = '/home/user/folder/project/src/style.scss';
    const expectedEnumName = 'Style';

    // Act
    const enumName = EnumGenerator.generateEnumNameFromPath(filePath);

    // Assert
    expect(enumName).toEqual(expectedEnumName);
  });

  it('should generate enum name from a provided path to a css stylesheet', () => {
    // Arrange
    const filePath = '/home/user/folder/project/src/style.css';
    const expectedEnumName = 'Style';

    // Act
    const enumName = EnumGenerator.generateEnumNameFromPath(filePath);

    // Assert
    expect(enumName).toEqual(expectedEnumName);
  });

  it('should generate enum name from a provided path to a stylesheet when stylesheet file name uses kebab casing', () => {
    // Arrange
    const filePath = '/home/user/folder/project/src/my-awesome-style-sheet.scss';
    const expectedEnumName = 'MyAwesomeStyleSheet';

    // Act
    const enumName = EnumGenerator.generateEnumNameFromPath(filePath);

    // Assert
    expect(enumName).toEqual(expectedEnumName);
  });

  it('should generate enum name from a provided path to a stylesheet when stylesheet file name uses snake casing', () => {
    // Arrange
    const filePath = '/home/user/folder/project/src/my_awesome_style_sheet.scss';
    const expectedEnumName = 'MyAwesomeStyleSheet';

    // Act
    const enumName = EnumGenerator.generateEnumNameFromPath(filePath);

    // Assert
    expect(enumName).toEqual(expectedEnumName);
  });

  it('should generate enum name from a provided path to a stylesheet when stylesheet file name uses dots to separate words', () => {
    // Arrange
    const filePath = '/home/user/folder/project/src/my-awesome.component.scss';
    const expectedEnumName = 'MyAwesomeComponent';

    // Act
    const enumName = EnumGenerator.generateEnumNameFromPath(filePath);

    // Assert
    expect(enumName).toEqual(expectedEnumName);
  });
});

describe(`${EnumGenerator.name} tests for enum body generation`, () => {
  it('should generate a string containing enum body with proper keys and values', () => {
    // Arrange
    const cssClasses = ['.class-one', '.class--two', '.class_three', '.class__four', '.class__five--item'];
    const expectedEnumPairs = [
      `ClassOne = 'class-one'`,
      `ClassTwo = 'class--two'`,
      `ClassThree = 'class_three'`,
      `ClassFour = 'class__four'`,
      `ClassFiveItem = 'class__five--item'`,
    ];

    // Act
    const enumBody = EnumGenerator.generateEnumBody(cssClasses);

    // Assert
    expect(expectedEnumPairs.every((i) => enumBody.includes(i))).toBeTruthy();
    expect(enumBody).toMatchSnapshot();
  });
});
