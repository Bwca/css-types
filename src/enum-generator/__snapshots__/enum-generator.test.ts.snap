// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EnumGenerator tests for enum body generation should generate a string containing enum body with proper keys and values 1`] = `
"{
  ClassOne = 'class-one',
  ClassTwo = 'class--two',
  ClassThree = 'class_three',
  ClassFour = 'class__four',
  ClassFiveItem = 'class__five--item',
}
"
`;
